### PR TITLE
Enhance nonunique()

### PR DESCRIPTION
**Pull request by alyst**
Use `DataFrameRow` as a working horse for `nonunique()` to significantly reduce memory usage and type conversions.

### CODE REVIEW

File: src/dataframerow/dataframerow.jl [hunk 1]
```diff
@@ -36,3 +36,36 @@ Base.next(r::DataFrameRow, s) = ((_names(r)[s], r[s]), s + 1)
...
+# compare two elements in the pooled array
+# NOTE assume there are no duplicated elements in the pool
+_isequal( a::PooledDataArray, i::Int, j::Int ) = isequal(a.refs[i], a.refs[j])
+
+function Base.isequal(r1::DataFrameRow, r2::DataFrameRow)
+    if r1.df !== r2.df throw( ArgumentError("Comparing rows from different frames not supported") ) end
+    if r1.row == r2.row return true end
+    for col in columns(r1.df)
+        if !_isequal(col, r1.row, r2.row) return false end
+    end
```
Comment (garborg) [inline src/dataframerow/dataframerow.jl]: Once we `NullableArray` replaces `DataArray`, we should replace `_isequal` with `isequal(col[row1], col[row2])`.

Comment (garborg) [inline src/dataframerow/dataframerow.jl]: Separate lines for `if`, `else`, `cond`, `end`, etc. -- I think basic style is outlined in CONTRIBUTING.md.

File: src/dataframerow/dataframerow.jl [hunk 2]
```diff
@@ -36,3 +36,36 @@ Base.next(r::DataFrameRow, s) = ((_names(r)[s], r[s]), s + 1)
...
+    return h
+end
+
+# compare two elements in the array
+_isequal( a::Array, i::Int, j::Int ) = isequal(a[i], a[j])
+
+# compare two elements in the data array
+function _isequal( a::DataArray, i::Int, j::Int )
+    if isna(a, i) return isna(a, j)
```
Comment (garborg) [inline src/dataframerow/dataframerow.jl]: No spaces inside parentheses.

File: src/dataframerow/dataframerow.jl [hunk 3]
```diff
@@ -36,3 +36,36 @@ Base.next(r::DataFrameRow, s) = ((_names(r)[s], r[s]), s + 1)
...
 
 Base.convert(::Type{Array}, r::DataFrameRow) = convert(Array, r.df[r.row,:])
+
+function Base.hash(r::DataFrameRow, h::UInt)
+    for col in columns(r.df)
+        if isna(col, r.row)
+            h = hash(false, h)
+        else
+            h = hash(true,hash(col[r.row], h))
+        end
```
Comment (garborg) [inline src/dataframerow/dataframerow.jl]: Spaces after commas.

**Comment by garborg** [comment]
This looks good to merge once it matches project style.

File: src/dataframerow/dataframerow.jl (hunk 1 above)
Comment (alyst) [inline src/dataframerow/dataframerow.jl]: >```isequal(col[row1], col[row2])```
But wouldn't that create an additional overhead for NullableArrays, where the elements would have to be converted into Nullable{T}? Also for PooledDataArrays (if they are gonna remain in some form) we can limit the comparison to value indexes only, if we know that they originate from the same pool.

**Comment by alyst** [comment]
@garborg Thanks for the review!
Actually, before merging this I also wanted to clarify if the logic of ```hash()``` and ```isequal()``` implemented here (i.e. matching by content, not by the row index and disallowing comparisons of rows from different frames) meets the original ```DataFrameRow``` concept.

**Comment by alyst** [comment]
I've fixed the code style, also added test for DataFrameRow operations. Tests pass, so I'm merging.

**Pull request merged**